### PR TITLE
Get image metadata from Airsim telemetry.

### PR DESCRIPTION
**Pull request by bluecamel**

### CODE REVIEW

File: ros2/src/airsim_ros_pkgs/include/math_common.h [hunk 1]
```diff
@@ -42,4 +42,13 @@ inline T angular_dist(T from, T to)
         d += 2. * M_PI;
     return d;
 }
+
+template <typename T>
+inline T constrain_euler_angle_to_360(const T angle)
+{
+    T angle_constrained = angle;
+    angle_constrained -= 360. * std::floor(angle_constrained * (1. / 360.));
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/include/math_common.h]: i think we can just fmod like T angle_constrained = std::fmod(angle, 360); here

File: ros2/src/airsim_ros_pkgs/include/math_common.h [hunk 2]
```diff
@@ -42,4 +42,13 @@ inline T angular_dist(T from, T to)
         d += 2. * M_PI;
     return d;
 }
+
+template <typename T>
+inline T constrain_euler_angle_to_360(const T angle)
+{
+    T angle_constrained = angle;
+    angle_constrained -= 360. * std::floor(angle_constrained * (1. / 360.));
+    return angle_constrained;
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/include/math_common.h]: may wanna handle negative cases for safety
so something like 

    if (angle_constrained < 0) {
        angle_constrained += 360;
    }

File: ros2/src/airsim_ros_pkgs/include/math_common.h [hunk 3]
```diff
@@ -42,4 +42,13 @@ inline T angular_dist(T from, T to)
...
+
+template <typename T>
+inline T constrain_euler_angle_to_360(const T angle)
+{
+    T angle_constrained = angle;
+    angle_constrained -= 360. * std::floor(angle_constrained * (1. / 360.));
+    return angle_constrained;
+}
+
 }
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/include/math_common.h:53]: I think this takes care of corners

template <typename T>
inline T constrain_euler_angle_to_360(const T angle) {
    T angle_constrained = std::fmod(angle, 360);
    if (angle_constrained < 0) {
        angle_constrained += 360;
    }
    return angle_constrained;
}

File: ros2/src/airsim_ros_pkgs/include/math_common.h (hunk 1 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/include/math_common.h]: lol, that's what I did initially, but 0 was sometimes 360, but I think I had an issue with my logging when I tested that.  I'll try it again.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 4]
```diff
@@ -1328,6 +1316,42 @@ void AirsimROSWrapper::lidar_timer_cb()
...
+    auto vehicle_attitude = get_camera_attitude_euler_msg(vehicle_name, "front_center_fixed");
+    auto gimbal_attitude = get_camera_attitude_euler_msg(vehicle_name, camera_name);
+
+    auto message = std::make_shared<airsim_interfaces::msg::GimbalAngleEulerCmd>();
+    message->header.stamp = nh_->get_clock()->now();
+    message->roll = gimbal_attitude->roll;
+    message->pitch = gimbal_attitude->pitch;
+
+    // Calculate "vehicle relative" yaw of the gimbaled camera.
+    message->yaw = math_common::constrain_euler_angle_to_360(gimbal_attitude->yaw - vehicle_attitude->yaw);
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1331]: can there be wrap around cases? unsure. do you think we should constraint_euler individually and then constraint the subtraction agian? I might be overthinking.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 5]
```diff
@@ -1404,6 +1429,48 @@ std::shared_ptr<sensor_msgs::msg::CompressedImage> AirsimROSWrapper::get_jpeg_ms
...
+            7, // num_glonass_satellites_used
+            11, // num_gps_satellites_used
+            30, // num_total_satellites_used
+            97.0, // position_dilution_precision
+            height_data.distance, // relative_height
+            1, // rtk_connection_status
+            {gps_data.gnss.geo_point.latitude, gps_data.gnss.geo_point.longitude,
+                gps_data.gnss.geo_point.altitude - origin_geo_point_.altitude}, // rtk_position
+            40, // rtk_position_info
+            {-0.326, -9.269, 152.085}, // rtk_velocity
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1463]: assuming this is just dummy values for now?

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 4 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1331]: They're already constrained above (in `get_camera_attitude_euler_msg`) before the subtraction and then again after.  I think that covers it?

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 6]
```diff
@@ -1404,6 +1429,48 @@ std::shared_ptr<sensor_msgs::msg::CompressedImage> AirsimROSWrapper::get_jpeg_ms
     return jpeg_message;
 }
 
+std::shared_ptr<image_metadata::Metadata> AirsimROSWrapper::get_jpeg_msg_metadata()
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1432]: is this the metadata being written into the images we will get for digital twin? in that case should we not write the 360 normalized data that is global? I are gimbal angles global here? as in, for example gimbal_attitude->yaw is from north?

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 7]
```diff
@@ -1328,6 +1316,42 @@ void AirsimROSWrapper::lidar_timer_cb()
...
+std::shared_ptr<airsim_interfaces::msg::GimbalAngleEulerCmd> AirsimROSWrapper::get_camera_attitude_euler_msg(
+    const std::string& vehicle_name, const std::string& camera_name)
+{
+    auto message = std::make_shared<airsim_interfaces::msg::GimbalAngleEulerCmd>();
+    message->header.stamp = nh_->get_clock()->now();
+
+    // Get the RPY orientation of a designated fixed camera (i.e. the vehicle orientation).
+    msr::airlib::CameraInfo vehicle_camera_info = airsim_client_->simGetCameraInfo(camera_name, vehicle_name);
+    tf2::Matrix3x3(get_tf2_quat(vehicle_camera_info.pose.orientation)).getRPY(
+        message->roll, message->pitch, message->yaw);
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1345]: is the camera_info.pose.orientation global?

File: ros2/src/airsim_ros_pkgs/include/math_common.h (hunk 1 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/include/math_common.h]: Okay, updated.

File: ros2/src/airsim_ros_pkgs/include/math_common.h (hunk 2 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/include/math_common.h]: Updated to account for negatives.

File: ros2/src/airsim_ros_pkgs/include/math_common.h (hunk 3 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/include/math_common.h:53]: Updated.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 5 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1463]: Yeah, I thought we decided that this doesn't need to be simulated.  Is that not correct?

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 4 above)
Comment (ahmadhasan2k8) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1331]: uh yeah, i missed it sorry.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 5 above)
Comment (ahmadhasan2k8) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1463]: yeah, right. just wanted to make sure

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 6 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1432]: Yes, this function is used by `get_jpeg_msg_from_img_msg` to get the metadata, which adds it to the image before publishing.

Vehicle and gimbal attitudes are global.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 7 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1345]: Yes.

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp [hunk 8]
```diff
@@ -1404,6 +1429,48 @@ std::shared_ptr<sensor_msgs::msg::CompressedImage> AirsimROSWrapper::get_jpeg_ms
...
 
+std::shared_ptr<image_metadata::Metadata> AirsimROSWrapper::get_jpeg_msg_metadata()
+{
+    std::string vehicle_name = "Copter";
+    auto height_data = airsim_client_->getDistanceSensorData("Distance1", vehicle_name);
+    auto distance_data = airsim_client_->getDistanceSensorData("Distance2", vehicle_name);
+    auto gps_data = airsim_client_->getGpsData("Gps", vehicle_name);
+
+    auto vehicle_attitude = get_camera_attitude_euler_msg(vehicle_name, "front_center_fixed");
+    auto gimbal_attitude = get_camera_attitude_euler_msg(vehicle_name, gimbal_vehicle_reference_camera_name_);
```
Comment (ahmadhasan2k8) [inline ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1440]: so, if i'm trying to create a global tf matrix, do i only use the gimbal attitude or gimbal and vehicle both?

**Review by ahmadhasan2k8 — COMMENTED** [review]
overall looks good, didn't get a chance to run yet. but if branton has and is good to go we can merge. do need some clarity on my end on gimbal angles being global vs relative in the metadata. as soon as I understand that, we can merge

**Comment by ahmadhasan2k8** [comment]
@bluecamel was commenting as i was reviewing, noticed late. good to go

**Review by ahmadhasan2k8 — APPROVED** [review]
looks good

File: ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp (hunk 8 above)
Comment (bluecamel) [reply ros2/src/airsim_ros_pkgs/src/airsim_ros_wrapper.cpp:1440]: I don't understand the question.  Both are global.

**Pull request merged**